### PR TITLE
infoschema: Add SERVER_CONNECT_ATTRS table

### LINKED ISSUE #44341: Support SESSION_CONNECT_ATTRS table
## Enhancement

In MySQL: https://dev.mysql.com/doc/refman/8.0/en/performance-schema-session-connect-attrs-table.html

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Issue Number: close #44341 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Connection Attributes can now be queried via the `INFORMATION_SCHEMA.SESSION_CONNECT_ATTRS` table.
```

### CODE REVIEW

**Comment by dveeden** [comment]
TiDB, with this PR:
```
sql> SELECT * FROM information_schema.SESSION_CONNECT_ATTRS;
+----------------+-----------------+------------+------------------+
| PROCESSLIST_ID | ATTR_NAME       | ATTR_VALUE | ORDINAL_POSITION |
+----------------+-----------------+------------+------------------+
|  2199023255955 | _client_name    | libmysql   |                0 |
|  2199023255955 | _client_version | 8.0.33     |                1 |
|  2199023255955 | _os             | Linux      |                2 |
|  2199023255955 | _pid            | 712927     |                3 |
|  2199023255955 | _platform       | x86_64     |                4 |
|  2199023255955 | program_name    | mysqlsh    |                5 |
+----------------+-----------------+------------+------------------+
6 rows in set (0.0012 sec)
```

MySQL 8.0.33
```
sql> SELECT * FROM performance_schema.session_connect_attrs;
+----------------+-----------------+------------+------------------+
| PROCESSLIST_ID | ATTR_NAME       | ATTR_VALUE | ORDINAL_POSITION |
+----------------+-----------------+------------+------------------+
|             20 | _pid            | 713027     |                0 |
|             20 | _platform       | x86_64     |                1 |
|             20 | _os             | Linux      |                2 |
|             20 | _client_name    | libmysql   |                3 |
|             20 | _client_version | 8.0.33     |                4 |
|             20 | program_name    | mysqlsh    |                5 |
+----------------+-----------------+------------+------------------+
6 rows in set (0.0039 sec)
```

Note the difference in `ORDINAL_POSITION` and `information_schema` vs `performance_schema`.

**Comment by dveeden** [comment]
Comparing TiDB (on port 4000) with MySQL (on port 3306):
```
[dvaneeden@dve-carbon tidb]$ mysqlsh 'mysql://root@127.0.0.1:4000/test?connection-attributes=[tidb=rocks,]' --table -e "TABLE performance_schema.session_connect_attrs"
+----------------+-----------------+------------+------------------+
| PROCESSLIST_ID | ATTR_NAME       | ATTR_VALUE | ORDINAL_POSITION |
+----------------+-----------------+------------+------------------+
|  2199023255961 | _client_name    | libmysql   |                0 |
|  2199023255961 | _client_version | 8.0.33     |                1 |
|  2199023255961 | _os             | Linux      |                2 |
|  2199023255961 | _pid            | 925769     |                3 |
|  2199023255961 | _platform       | x86_64     |                4 |
|  2199023255961 | program_name    | mysqlsh    |                5 |
|  2199023255961 | tidb            | rocks      |                6 |
+----------------+-----------------+------------+------------------+
[dvaneeden@dve-carbon tidb]$ mysqlsh 'mysql://root:root@127.0.0.1/test?connection-attributes=[tidb=rocks,]' --table -e "TABLE performance_schema.session_connect_attrs"
+----------------+-----------------+------------+------------------+
| PROCESSLIST_ID | ATTR_NAME       | ATTR_VALUE | ORDINAL_POSITION |
+----------------+-----------------+------------+------------------+
|             32 | _platform       | x86_64     |                0 |
|             32 | _os             | Linux      |                1 |
|             32 | _client_name    | libmysql   |                2 |
|             32 | _pid            | 925806     |                3 |
|             32 | tidb            | rocks      |                4 |
|             32 | _client_version | 8.0.33     |                5 |
|             32 | program_name    | mysqlsh    |                6 |
+----------------+-----------------+------------+------------------+
```

**Review by tiancaiamao — APPROVED** [review]

File: infoschema/perfschema/const.go [hunk 1]
```diff
@@ -549,3 +550,10 @@ const tablePDProfileGoroutines = "CREATE TABLE IF NOT EXISTS " + tableNamePDProf
 const tableSessionVariables = "CREATE TABLE IF NOT EXISTS " + tableNameSessionVariables + " (" +
 	"VARIABLE_NAME VARCHAR(64) NOT NULL," +
 	"VARIABLE_VALUE VARCHAR(1024) NOT NULL);"
+
+// tableSessionConnectAttrs contains the column name definitions for the table session_connect_attrs
+const tableSessionConnectAttrs = "CREATE TABLE IF NOT EXISTS " + tableNameSessionConnectAttrs + " (" +
+	"PROCESSLIST_ID bigint unsigned NOT NULL," +
+	"ATTR_NAME varchar(32) COLLATE utf8mb4_bin NOT NULL," +
+	"ATTR_VALUE varchar(1024) COLLATE utf8mb4_bin DEFAULT NULL," +
+	"ORDINAL_POSITION int DEFAULT NULL);"
```
Comment (Benjamin2037) [inline infoschema/perfschema/const.go:559]: I notice MySQL [session_connect_attrs](https://dev.mysql.com/doc/refman/8.0/en/performance-schema-session-connect-attrs-table.html) has definied PK. why we here do not need PK for the table?

File: infoschema/tables.go [hunk 2]
```diff
@@ -2003,6 +2004,37 @@ func GetDataFromSessionVariables(ctx context.Context, sctx sessionctx.Context) (
...
 }
 
+// GetDataFromSessionConnectAttrs produces the rows for the session_connect_attrs table.
+func GetDataFromSessionConnectAttrs(sctx sessionctx.Context) ([][]types.Datum, error) {
+	sm := sctx.GetSessionManager()
+	if sm == nil {
+		return nil, nil
+	}
+	allAttrs := sm.GetConAttrs()
+	rows := make([][]types.Datum, 0, len(allAttrs)*10) // 10 Attributes per connection
```
Comment (Benjamin2037) [inline infoschema/tables.go:2014]: Why here set 10 Attributes? Can we use actual Attributes number column?

File: infoschema/perfschema/const.go (hunk 1 above)
Comment (dveeden) [reply infoschema/perfschema/const.go:559]: It looks like none of the I_S and P_S tables have a primary key, so I followed that. Not sure if that's required or not. I also don't think having a PK makes a big difference here?

```
sql> select * from information_schema.KEY_COLUMN_USAGE WHERE CONSTRAINT_NAME='PRIMARY' AND table_schema='performance_schema';
Empty set (0.0048 sec)

sql> select * from information_schema.KEY_COLUMN_USAGE WHERE CONSTRAINT_NAME='PRIMARY' AND table_schema='information_schema';
Empty set (0.0042 sec)
```

File: infoschema/tables.go (hunk 2 above)
Comment (dveeden) [reply infoschema/tables.go:2014]: As `allAttrs` is a map of maps we would need to loop over all maps and get the sum of the lengths of all maps. Every map is one connection and often there are about 6 attributes per connection. So getting the exact map takes more CPU time while the less exact number might allocate more rows than needed, or would under allocate if there are many rows per connection.

```go
package main

import "testing"

func setup() map[uint64]map[string]string {
	m := make(map[uint64]map[string]string)
	for i := uint64(0); i < 100000; i++ {
		m[i] = map[string]string{
			"foo1": "bar",
			"foo2": "bar",
			"foo3": "bar",
			"foo4": "bar",
			"foo5": "bar",
			"foo6": "bar",
		}
	}
	return m
}

func BenchmarkSimple(b *testing.B) {
	m := setup()
	b.ResetTimer()
	println("Simple length of m: ", len(m)*10)
}

func BenchmarkExact(b *testing.B) {
	m := setup()
	b.ResetTimer()
	var t int
	for _, k := range m {
		t = t + len(k)
	}
	println("Exact length of m: ", t)
}
```

```
$ go test -bench .
Simple length of m:  1000000
goos: linux
goarch: amd64
pkg: foo
cpu: Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz
BenchmarkSimple-8   	Simple length of m:  1000000
Simple length of m:  1000000
Simple length of m:  1000000
Simple length of m:  1000000
Simple length of m:  1000000
1000000000	         0.0000109 ns/op
Exact length of m:  600000
BenchmarkExact-8    	Exact length of m:  600000
Exact length of m:  600000
Exact length of m:  600000
Exact length of m:  600000
Exact length of m:  600000
1000000000	         0.003883 ns/op
PASS
ok  	foo	0.625s
```

Comment (dveeden) [reply infoschema/tables.go:2014]: I think both methods are fine. Should I change the code to use the more exact one?

File: infoschema/perfschema/const.go (hunk 1 above)
Comment (Benjamin2037) [reply infoschema/perfschema/const.go:559]: if there is a little rows in this table, it is no big difference.

**Pull request merged**